src/ripple/basics/impl/PerfLogImp.h

```diff
@@ -0,0 +1,250 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2018 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_BASICS_PERFLOGIMP_H
+#define RIPPLE_BASICS_PERFLOGIMP_H
+
+#include <ripple/basics/PerfLog.h>
+#include <ripple/basics/chrono.h>
+#include <ripple/core/Stoppable.h>
+#include <ripple/app/main/Application.h>
+#include <ripple/protocol/JsonFields.h>
+#include <ripple/beast/core/SystemStats.h>
+#include <cstdint>
+#include <mutex>
+#include <memory>
+#include <unordered_map>
+#include <string>
+#include <utility>
+#include <tuple>
+#include <stdexcept>
+#include <vector>
+#include <fstream>
+#include <thread>
+#include <condition_variable>
+
+namespace ripple {
+namespace perf {
+
+/**
+ * Implementation class for PerfLog.
+ */
+class PerfLogImp
+    : public PerfLog, Stoppable
+{
+    /**
+     * Track performance counters and currently executing tasks.
+     */
+    class Counters
+    {
+    public:
+        /**
+         * RPC performance counters.
+         */
+        struct Rpc
+        {
+            std::uint64_t started {0};
+            std::uint64_t finished {0};
+            std::uint64_t errored {0};
```

Comment:
@scottschurr How to render those then? Convert a time_point to a duration, then do count()?

Reply:
I used this, which is exactly what you described:
```
std::to_string((proc.second.started - steady_time_point{}).count());
```
That gives you the `count` since `steady_clock`'s epoch, which I think is what you want.  But you leave the conversion until you absolutely have to do it.  Up to that point you're working in `chrono` units.  That's what give you the type safety.

Reply:
@scottschurr No, those are counters. I'm changing the duration to be microseconds though. I'm putting comments in the structs to describe what the members do.